src/index.ts

```diff
@@ -22,37 +52,73 @@ export const goSync = <T, E extends Error>(fn: () => T): GoResult<T, E> => {
   }
 };
 
-const createGoError = <E extends Error>(err: unknown): GoResultError<E> => {
-  if (err instanceof Error) return fail(err);
-  return fail(new Error('' + err));
+const retryFnWrapper = async <T, E extends Error>(
+  fn: Promise<T> | (() => Promise<T>),
+  attemptOptions: AttemptOptions<T>
+): Promise<GoResult<T, E>> => {
+  if (typeof fn === 'function') {
+    return retryFn(fn, attemptOptions);
+  }
+  return retryFn(() => fn, attemptOptions);
+};
+
+const retryFn = async <T, E extends Error>(
+  fn: () => Promise<T>,
+  attemptOptions: AttemptOptions<T>
+): Promise<GoResult<T, E>> => {
+  return retry(fn, attemptOptions)
+    .then(success)
+    .catch((err) => {
+      return createGoError(err);
+    });
 };
 
-export const go = async <T, E extends Error>(fn: Promise<T> | (() => Promise<T>)): Promise<GoResult<T, E>> => {
+export const go = async <T, E extends Error>(
+  fn: Promise<T> | (() => Promise<T>),
+  options?: PromiseOptions
+): Promise<GoResult<T, E>> => {
+  const attemptOptions: AttemptOptions<any> = {
+    delay: options?.retryDelayMs || 200,
+    maxAttempts: (options?.retries || 0) + 1,
+    initialDelay: 0,
+    minDelay: 0,
+    maxDelay: 0,
+    factor: 0,
+    timeout: options?.timeoutMs || 0,
+    jitter: false,
+    handleError: null,
+    handleTimeout: options?.timeoutMs
+      ? async (context: AttemptContext, options: AttemptOptions<T>) => {
+          if (context.attemptsRemaining > 0) {
+            const res = await retryFnWrapper(fn, { ...options, maxAttempts: context.attemptsRemaining });
+
+            if (res.success) {
+              return res.data;
+            } else {
+              throw res.error;
+            }
+          }
+          throw new Error(`Operation timed out after final retry`);
```

Comment:
This error message will also be used when no retries are used so it's a bit connfusing. Let's just change that to `Operation timed out`.